### PR TITLE
[export] Support preserving calling convention to some modules.

### PR DESCRIPTION
**Pull request by zhxchen17**
Summary: APS use this feature to swap out some submodules after unflattening.

Test Plan: test_export_preserve_signature

Differential Revision: D48154341

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_export/wrappers.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        def __init__(self, inner):
+            super().__init__()
+            self.inner = inner
+            self.module_call_signatures = module_call_signatures
+
+        def forward(self, *args, **kwargs):
+            flat_args, in_spec = pytree.tree_flatten((args, kwargs))
+
+            def flat_gm(*flat_args):
+                flat_args = export_tracepoint(*flat_args, kind="module_call_in", path=path)
```
Comment (tugsbayasgalan) [inline torch/_export/wrappers.py:91]: nit: Can you rename this to module_inputs?

File: torch/_export/wrappers.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                args, kwargs = pytree.tree_unflatten(flat_args, in_spec)
+                res = self.inner(*args, **kwargs)
+                flat_res, out_spec = pytree.tree_flatten(res)
+                flat_res = export_tracepoint(*flat_res, kind="module_call_out", path=path)
```
Comment (tugsbayasgalan) [inline torch/_export/wrappers.py:95]: nit: module_outputs

File: torch/_export/passes/collect_tracepoints_pass.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.specs = specs
+
+    def call(self, gm):
+        for module in gm.modules():
+            if not isinstance(module, torch.fx.GraphModule):
+                continue
+            for node in module.graph.nodes:
+                if node.op != "call_function":
+                    continue
+                if node.target == torch.ops.higher_order.export_tracepoint:
```
Comment (tugsbayasgalan) [inline torch/_export/passes/collect_tracepoints_pass.py:27]: This kind of implies that export_tracepoint is public?

File: torch/_export/passes/collect_tracepoints_pass.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                    for i, arg in enumerate(node.args):
+                        kind = node.kwargs["kind"]
+                        if kind == "module_call_in":
+                            self.specs[node.kwargs['path']].inputs.append(arg.name)
+                        elif kind == "module_call_out":
+                            self.specs[node.kwargs['path']].outputs.append(arg.name)
+                        else:
+                            assert False
+                        for user in node.users:
+                            assert user.op == "call_function"
```
Comment (tugsbayasgalan) [inline torch/_export/passes/collect_tracepoints_pass.py:37]: Hmm why are they getattrs?

File: torch/_export/passes/collect_tracepoints_pass.py (hunk 3 above)
Comment (zhxchen17) [reply torch/_export/passes/collect_tracepoints_pass.py:27]: ok I will add a _

**Review by tugsbayasgalan — APPROVED** [review]

**Comment by zejun-chen** [comment]
Hi, @zhxchen17 

May i know what does the `APS` mean?
Does it mean the user can use this higher order op to preserve something in `torch.export`?

Thank you.

**Pull request closed without merging**